drivers/net/wireless/marvell/mwifiex/pcie.c

```diff
@@ -665,20 +665,50 @@ static int mwifiex_pm_wakeup_card(struct mwifiex_adapter *adapter)
 {
 	struct pcie_service_card *card = adapter->card;
 	const struct mwifiex_pcie_card_reg *reg = card->pcie.reg;
+	int n_tries;
 
 	mwifiex_dbg(adapter, EVENT,
 		    "event: Wakeup device...\n");
 
 	if (reg->sleep_cookie)
 		mwifiex_pcie_dev_wakeup_delay(adapter);
 
-	/* Accessing fw_status register will wakeup device */
+	/* Accessing fw_status register will wakeup device.
+	 * Since the 88W8897 firmware sometimes appears to ignore or miss
+	 * that wakeup request, we continue trying down below until we
+	 * receive an interrupt from the card.
+	 */
 	if (mwifiex_write_reg(adapter, reg->fw_status, FIRMWARE_READY_PCIE)) {
 		mwifiex_dbg(adapter, ERROR,
 			    "Writing fw_status register failed\n");
 		return -1;
 	}
 
+	n_tries = 1;
+
+	usleep_range(400, 700);
+
+	/* adapter->int_status is a single byte, so should be fine to access
+	 * without aquiring the spin-lock.
+	 */
+	while (n_tries < 50 && adapter->int_status == 0) {
+		if (mwifiex_write_reg(adapter, reg->fw_status, FIRMWARE_READY_PCIE)) {
+			mwifiex_dbg(adapter, ERROR,
+				   "Writing fw_status register failed\n");
+			return -1;
+		}
+
+		if (n_tries < 15)
+			usleep_range(400, 700);
+		else
+			msleep(10);
+
+		n_tries++;
+	}
```

Comment:
I think you could move all of this into a single do-while loop. Initialize `n_tries` above to zero and then

```c
do {
	if (n_tries >= 15)
		msleep(10);
	else if (n_tries)
		usleep_range(400, 700);

	if (mwifiex_write_reg(adapter, reg->fw_status, FIRMWARE_READY_PCIE)) {
		mwifiex_dbg(adapter, ERROR, "Writing fw_status register failed\n");
		return -1;
	}

	n_tries++;
} while (n_tries < 50 && adapter->int_status == 0)
```

would make this a bit more compact.

Reply:
Good idea! Did something similar to this